x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyAuthenticator.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.authc;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.xpack.core.security.authc.Authentication;
+import org.elasticsearch.xpack.core.security.authc.AuthenticationResult;
+import org.elasticsearch.xpack.core.security.authc.AuthenticationToken;
+import org.elasticsearch.xpack.core.security.support.Exceptions;
+
+class ApiKeyAuthenticator implements Authenticator {
+
+    private static final Logger logger = LogManager.getLogger(ApiKeyAuthenticator.class);
+
+    private final ApiKeyService apiKeyService;
+    private final String nodeName;
+
+    ApiKeyAuthenticator(ApiKeyService apiKeyService, String nodeName) {
+        this.apiKeyService = apiKeyService;
+        this.nodeName = nodeName;
+    }
+
+    @Override
+    public String name() {
+        return "API key";
+    }
+
+    @Override
+    public AuthenticationToken extractCredentials(Context context) {
+        return apiKeyService.getCredentialsFromHeader(context.getThreadContext());
+    }
+
+    @Override
+    public void authenticate(Context context, ActionListener<Result> listener) {
+        final AuthenticationToken authenticationToken = context.getAuthenticationToken();
+        if (false == authenticationToken instanceof ApiKeyService.ApiKeyCredentials) {
+            listener.onResponse(Authenticator.Result.notHandled());
+            return;
+        }
+        ApiKeyService.ApiKeyCredentials apiKeyCredentials = (ApiKeyService.ApiKeyCredentials) authenticationToken;
+        apiKeyService.tryAuthenticate(context.getThreadContext(), apiKeyCredentials, ActionListener.wrap(authResult -> {
+            if (authResult.isAuthenticated()) {
+                final Authentication authentication = apiKeyService.createApiKeyAuthentication(authResult, nodeName);
+                listener.onResponse(Authenticator.Result.success(authentication));
+            } else if (authResult.getStatus() == AuthenticationResult.Status.TERMINATE) {
+                Exception e = (authResult.getException() != null) ?
+                    authResult.getException() :
+                    Exceptions.authenticationError(authResult.getMessage());
+                logger.debug(new ParameterizedMessage("API key service terminated authentication for request [{}]", context.getRequest()),
+                    e);
+                listener.onFailure(e);
```

Comment:
My intention was the opposite. That is my plan was to also remove `TERMINATE` from `AuthenticationResult` because I failed to see its usefulness compare to just call `onFailure`. I started with having a `TERMINATE` for `Result` as well, but realised its handling was basically just to call `onFailure` before logging a message. I also found it confusing that `AuthenticationResult` has a single `CONTINUE` status to represent:

1. The realm/service didn not handle the authentication; and
2. The realm/service cannot authenticate the request, e.g. invalid credentials, but also decide it is not fatal 

That is why for the new `Result`, I removed `TERMINATE` and replaced `CONTINUE` with `NOT_HANDLED` and `UNSUCCESSFUL`. 

In theory, we differentiate `onFailure` from `TERMINATE` for errors not really related to authentication itself, e.g. index not available etc. But in practice, there are so few usage of `TERMINATE` and they can be replaced by `onFailure`. If we do still need differentiate it, I think a Named exception is probably a better approach?

(I remember there was an issue with having `TERMINATE` complicates the handling of IteractingActionListener when it walks through the authenticators. But I cannot seem to remember it now. Maybe it is no longer an issue since I have changed the implementation a few times before raising the PR.)